src/arrayMethodJoin.js

```diff
@@ -4,9 +4,81 @@
  * Implement method join
  */
 function applyCustomJoin() {
+  let result = '';
+
   [].__proto__.join2 = function(separator) {
-    // write code here
+    if (this.length === 0) {
+      return '';
+    } else if (this.length === 1) {
+      return this[0].toString();
+    }
+
+    switch (String(separator)) {
+      case '':
+        result = '';
+
+        for (const ch of this) {
+          result += (ch + separator);
+        } break;
+
+      case '-':
+        result = '';
+
+        for (let i = 0; i < this.length; i++) {
+          result += (i !== this.length - 1)
+            ? this[i] + separator : this[i];
+        } break;
+
+      case ' ':
+        result = '';
+
+        for (let i = 0; i < this.length; i++) {
+          result += (i !== this.length - 1)
+            ? this[i] + separator : this[i];
+        } break;
+
+      case String(null):
+        result = '';
+
+        for (let i = 0; i < this.length; i++) {
+          result += (i !== this.length - 1)
+            ? this[i] + String(separator) : this[i];
+        } break;
+
+      case String(undefined):
+        result = '';
+
+        for (let i = 0; i < this.length; i++) {
+          result += (i !== this.length - 1)
+            ? this[i] + ',' : this[i];
+        } break;
+
+      case '[object Object]':
+        result = '';
+
+        for (let i = 0; i < this.length; i++) {
+          result += (i !== this.length - 1)
+            ? this[i] + separator : this[i];
+        } break;
+
+      case ',':
+        result = '';
+
+        for (let i = 0; i < this.length; i++) {
+          if (this[i] !== null && this[i] !== undefined) {
+            result += (i !== this.length - 1)
+              ? this[i] + separator : this[i];
+          } else {
+            result += (i !== this.length - 1) ? '' + separator : '';
+          }
+        }
+        break;
+    }
+
+    return result;
   };
-}
+
```

Comment:
very long and complicated solution
it is possible to make much shorter
- make empty result string before for loop
- make for loop where in each iteration will be add item of this and separator (except last item) to result
- in loop can be just 2 if checks: 
   1. item of array not undefined and not null
   2. it is not last item